io.openems.edge.timeofusetariff.entsoe/src/io/openems/edge/timeofusetariff/entsoe/TouEntsoeImpl.java

```diff
@@ -0,0 +1,188 @@
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.metatype.annotations.Designate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xml.sax.SAXException;
+
+import com.google.common.collect.ImmutableSortedMap;
+
+import io.openems.common.exceptions.OpenemsError.OpenemsNamedException;
+import io.openems.common.utils.ThreadPoolUtils;
+import io.openems.edge.common.channel.value.Value;
+import io.openems.edge.common.component.AbstractOpenemsComponent;
+import io.openems.edge.common.component.OpenemsComponent;
+import io.openems.edge.common.currency.Currency;
+import io.openems.edge.common.meta.Meta;
+import io.openems.edge.timeofusetariff.api.TimeOfUsePrices;
+import io.openems.edge.timeofusetariff.api.TimeOfUseTariff;
+import io.openems.edge.timeofusetariff.api.utils.TimeOfUseTariffUtils;
+
+@Designate(ocd = Config.class, factory = true)
+@Component(//
+		name = "TimeOfUseTariff.ENTSO-E", //
+		immediate = true, //
+		configurationPolicy = ConfigurationPolicy.REQUIRE //
+)
+public class TouEntsoeImpl extends AbstractOpenemsComponent implements TouEntsoe, OpenemsComponent, TimeOfUseTariff {
+
+	private final Logger log = LoggerFactory.getLogger(TouEntsoeImpl.class);
+	private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+	private final AtomicReference<ImmutableSortedMap<ZonedDateTime, Float>> prices = new AtomicReference<>(
+			ImmutableSortedMap.of());
+
+	private Config config = null;
+	private Currency currency;
+	private ZonedDateTime updateTimeStamp = null;
+	private static final int EUR_EXHANGE_RATE = 1;
+
+	@Reference
+	private CurrencyProvider currencyProvider;
+
+	public TouEntsoeImpl() {
+		super(//
+				OpenemsComponent.ChannelId.values(), //
+				TouEntsoe.ChannelId.values() //
+		);
+	}
+
+	private final Consumer<Value<Integer>> callback = t -> {
+		this.currency = t.asEnum();
+		this.executor.schedule(this.task, 0, TimeUnit.SECONDS);
+	};
+
+	@Activate
+	private void activate(ComponentContext context, Config config) {
+		super.activate(context, config.id(), config.alias(), config.enabled());
+
+		if (!config.enabled()) {
+			return;
+		}
+		this.config = config;
+		this.currency = this.currencyProvider.getCurrent();
+		this.executor.schedule(this.task, 0, TimeUnit.SECONDS);
+
+		// Trigger to activate when the currency from Meta is updated.
+		this.currencyProvider.subscribe(this.callback);
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		super.deactivate();
+		this.currencyProvider.unsubscribe(this.callback);
+		ThreadPoolUtils.shutdownAndAwaitTermination(this.executor, 0);
+	}
+
+	@Component
+	public static class CurrencyProviderOsgi implements CurrencyProvider {
```

Comment:
This is overcomplicating things in my opinion - please implement it the way we discussed.

Yes, introducing a `CurrencyProvider` could be feasible if in future we wanted to extract this feature out of `Core.Meta`. But to keep things simple - we can just refactor in that case if really thing it makes sense (which I don't think currently).

Reply:
I have implemented this way because of the TouEntsoeTest. In the TouEntsoeTest, the meta is null. One way would be to implement DummyMeta and handle the currency channel. So I thought this would be simple.